packages/sysutils/busybox/scripts/init

```diff
@@ -549,38 +552,110 @@
+    if [ -f "$UPDATE_DIR/$UPDATE_KERNEL" -a -f "$UPDATE_DIR/$UPDATE_SYSTEM" ] || [ -f "$UPDATE_TAR" -o -f "$UPDATE_IMG_GZ" -o -f "$UPDATE_IMG" ] ; then
       if [ "$UPDATE_DISABLED" = "yes" ] ; then
-        rm -rf $UPDATE_DIR/[0-9a-zA-Z]* &>/dev/null
         echo "Updating not supported on netboot. normal startup in 10s..."
+        do_cleanup
         sync
         usleep 10000000
         return 0
       fi
 
-      # check for .tar
+      [ -f "$UPDATE_DIR/.nocheck" ] && MD5_NOCHECK="1"
+      rm -fr "$UPDATE_DIR/.tmp" &>/dev/null
+
       if [ -f "$UPDATE_TAR" ] ; then
         echo "Found new .tar archive. extracting..."
         mkdir -p $UPDATE_DIR/.tmp &>/dev/null
         tar -xf "$UPDATE_TAR" -C $UPDATE_DIR/.tmp &>/dev/null
         mv $UPDATE_DIR/.tmp/*/target/* $UPDATE_DIR &>/dev/null
-        rm -f "$UPDATE_TAR" &>/dev/null
-        rm -rf $UPDATE_DIR/.tmp &>/dev/null
+      elif [ -f "$UPDATE_IMG_GZ" -o -f "$UPDATE_IMG" ] ; then
+        mkdir -p $UPDATE_DIR/.tmp/mnt &>/dev/null
+        IMG_FILE="$UPDATE_DIR/.tmp/update.img"
+
+        if [ -f "$UPDATE_IMG_GZ" ]; then
+          echo "Found new .img.gz archive. extracting..."
+          gunzip -d -c "$UPDATE_IMG_GZ" > $IMG_FILE
+        else
+          echo "Found new .img file..."
+          mv "$UPDATE_IMG" $IMG_FILE
+        fi
+
+        LOOP=$(losetup -f)
+        LOOP_NUM=$(echo $LOOP | sed 's|/dev/loop||')
+        mknod $LOOP b 7 $LOOP_NUM
+        losetup $LOOP $IMG_FILE
+
+        # check for MBR partititon
+        OFFSET=$(fdisk -u -l $LOOP | awk '/^[ ]*Device/{part=1; next}; part{if ($2 == "*") {print $3} else {print $2} ; exit}')
+        if [ -z "$OFFSET" ]; then
+          # check for GPT partititon
+          OFFSET=$(fdisk -u -l $LOOP | awk '/^Number/{part=1; next}; part{print $2; exit}')
+          if [ -z "$OFFSET" ]; then
+            do_cleanup
+            error "img update" "Can't get system partition from image file"
+          fi
+        fi
+
+        SECTOR_SIZE=$(cat /sys/devices/virtual/block/loop${LOOP_NUM}/queue/hw_sector_size)
+        losetup -d $LOOP
         sync
-        if [ ! -f "$UPDATE_DIR/$UPDATE_KERNEL" -o ! -f "$UPDATE_DIR/$UPDATE_SYSTEM" ] ; then
-          echo "missing ${UPDATE_KERNEL} or ${UPDATE_SYSTEM}... normal startup in 10s"
-          sync
-          usleep 10000000
-          return 0
+
+        OFFSET=$(($OFFSET * $SECTOR_SIZE))
+
+        # use losetup because busybox mount does
+        # not support the -o offset option
+        # also busybox umount deletes
+        # loop device automatically
+        echo "Mounting system partition..."
+        losetup -o $OFFSET $LOOP $IMG_FILE
+        mount -o ro,loop $LOOP $UPDATE_DIR/.tmp/mnt
+
+        # don't make temporary files but instead copy
+        # directly from mountpoint to /flash
+        UPDATE_DIR=$UPDATE_ROOT/.tmp/mnt
+        UPDATE_KERNEL="$IMAGE_KERNEL"
+
+        if [ ! -f "$UPDATE_DIR/${UPDATE_KERNEL}.md5" -o ! -f "$UPDATE_DIR/${UPDATE_SYSTEM}.md5" ]; then
```

Comment:
This is for discussion. Yes, checking md5 is the only check to prevent bad files. Maybe this decision should be removed (again).